File: eng-backend/src/main/java/com/nemisolv/starter/entity/Course.java
package com.nemisolv.starter.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Course {
    private Long id;
    private Long categoryId;
    private String title;
    private String slug;
    private String description;
    private String thumbnail;
    private DifficultyLevel difficultyLevel;
    private Integer estimatedDuration;
    private Integer totalLessons;
    private Integer totalExercises;
    private List<String> prerequisites;
    private List<String> learningObjectives;
    private List<String> tags;
    private Boolean isPublished;
    private Boolean isPremium;
    private BigDecimal price;
    private Long createdBy;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    // Category details (for joined queries)
    private String categoryName;
    private String categorySlug;

    public enum DifficultyLevel {
        BEGINNER,
        ELEMENTARY,
        INTERMEDIATE,
        UPPER_INTERMEDIATE,
        ADVANCED,
        PROFICIENT
    }
}

File: front-end/types/course.ts
export enum DifficultyLevel {
  BEGINNER = "BEGINNER",
  ELEMENTARY = "ELEMENTARY",
  INTERMEDIATE = "INTERMEDIATE",
  UPPER_INTERMEDIATE = "UPPER_INTERMEDIATE",
  ADVANCED = "ADVANCED",
  PROFICIENT = "PROFICIENT",
}

export enum LessonType {
  VIDEO = "VIDEO",
  AUDIO = "AUDIO",
  TEXT = "TEXT",
  INTERACTIVE = "INTERACTIVE",
  QUIZ = "QUIZ",
  EXERCISE = "EXERCISE",
}

export interface Course {
  id: number;
  categoryId?: number;
  categoryName?: string;
  categorySlug?: string;
  title: string;
  slug: string;
  description?: string;
  thumbnail?: string;
  difficultyLevel: DifficultyLevel;
  estimatedDuration?: number;
  totalLessons?: number;
  totalExercises?: number;
  prerequisites?: string[];
  learningObjectives?: string[];
  tags?: string[];
  isPublished: boolean;
  isPremium: boolean;
  price?: number;
  createdBy?: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface CourseRequest {
  title: string;
  slug?: string;
  description?: string;
  thumbnail?: string;
  categoryId?: number;
  difficultyLevel: DifficultyLevel;
  estimatedDuration?: number;
  prerequisites?: string[];
  learningObjectives?: string[];
  tags?: string[];
  isPublished?: boolean;
  isPremium?: boolean;
  price?: number;
}

export interface Lesson {
  id: number;
  courseId: number;
  courseTitle?: string;
  courseSlug?: string;
  title: string;
  slug: string;
  description?: string;
  content?: string;
  videoUrl?: string;
  audioUrl?: string;
  transcript?: string;
  duration?: number;
  lessonType: LessonType;
  sortOrder: number;
  isPublished: boolean;
  isPreview: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export interface LessonRequest {
  title: string;
  slug?: string;
  description?: string;
  content?: string;
  videoUrl?: string;
  audioUrl?: string;
  transcript?: string;
  duration?: number;
  lessonType: LessonType;
  sortOrder?: number;
  isPublished?: boolean;
  isPreview?: boolean;
}

export interface Category {
  id: number;
  name: string;
  slug: string;
  description?: string;
  icon?: string;
  color?: string;
  sortOrder: number;
  isActive: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export interface PaginationParams {
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: "ASC" | "DESC";
}

export interface CourseFilters extends PaginationParams {
  categoryId?: number;
  difficultyLevel?: DifficultyLevel;
  isPublished?: boolean;
  isPremium?: boolean;
  search?: string;
}

export interface LessonFilters {
  lessonType?: LessonType;
  isPublished?: boolean;
}
